src/kv/ZetaScaleStore.cc

```diff
@@ -0,0 +1,1030 @@
+
+  if (dev_log_fd >= 0) {
+    r = fdatasync(dev_log_fd);
+    if (r < 0)
+      derr << "fdatasync failed fd=" << dev_log_fd << " :" << strerror(errno)
+	   << dendl;
+  }
+
+  return status == ZS_SUCCESS ? 0 : -1;
+}
+
+int ZSStore::transaction_rollback()
+{
+  ZS_status_t status = ZSTransactionCommit(_thd_state());
+
+  if (status != ZS_SUCCESS)
+    derr << "ZSTransactionCommit failed: " << ZSStrError(status) << dendl;
+
+  return status == ZS_SUCCESS ? 0 : -1;
+}
+
+bool ZSStore::key_is_prefixed(const string &prefix, const string &full_key)
+{
+  if ((full_key.length() > prefix.length()) &&
+      (full_key.c_str()[prefix.length()] == '\0'))
+    return !memcmp(full_key.c_str(), prefix.c_str(), prefix.length());
+
+  return false;
+}
+
+void ZSStore::ZSTransactionImpl::set(const string &prefix, const string &k,
+				     const bufferlist &to_set_bl)
+{
+  dtrace << " " << prefix << " " << decode_key(k) << dendl;
+
+  ops.push_back(
+      make_pair(WRITE, make_pair(combine_strings(prefix, k, true), to_set_bl)));
+}
+
+void ZSStore::ZSTransactionImpl::rmkey(const string &prefix, const string &k)
+{
+  dtrace << " " << prefix << " " << decode_key(k) << dendl;
+
+  ops.push_back(make_pair(
+      DELETE, make_pair(combine_strings(prefix, k, true), bufferlist())));
+}
+
+void ZSStore::ZSTransactionImpl::rmkeys_by_prefix(const string &prefix)
+{
+  KeyValueDB::Iterator it = db->get_iterator(prefix);
+  for (it->seek_to_first(); it->valid(); it->next()) rmkey(prefix, it->key());
+}
+
+bufferlist ZSStore::_merge(const std::string &key, const bufferlist &_base,
+			   const bufferlist &value)
+{
+  bufferlist rbase;
+  std::string out;
+
+  if (!_base.length()) {
+    int r = _get(key, &rbase);
+    assert(!r || r == -ENOENT);
+  }
+
+  /* Use local, read from disk, value, when argument _base is empty
+   * Make _base contiguous if it's not */
+  const bufferlist &base =
+      !_base.length() ? rbase
+		      : (_base.is_contiguous() ? _base : bufferlist(_base));
+
+  /* Make "value" contiguous, so we can use c_str() of front buffer later */
+  const bufferlist &v =
+      value.is_contiguous() && value.length() > 0 ? value : bufferlist(value);
+
+  dtrace << " " << key << " " << base.length() << " " << value.length() << " "
+       << dendl;
+
+  for (auto &p : merge_ops) {
+    if (!key_is_prefixed(p.first, key))
+      continue;
+
+    if (base.length())
+      p.second->merge(base.front().c_str(), base.length(),
+		      v.buffers().front().c_str(), v.length(), &out);
+    else
+      p.second->merge_nonexistent(v.buffers().front().c_str(), v.length(),
+				  &out);
+  }
+
+  bufferlist res;
+  res.append(bufferptr(out.c_str(), out.length()));
+  return res;
+}
+
+void ZSStore::ZSTransactionImpl::merge(const std::string &prefix,
+				       const std::string &k,
+				       const bufferlist &value)
+{
+  ops.push_back(
+      make_pair(MERGE, make_pair(combine_strings(prefix, k, true), value)));
+}
+
+#define WRITE_BATCH_SIZE 256
```

Comment:
This looks suspicious... it looks like once you get 256 pending k/v updates you do a put.  Is this actually keeping the overall Transaction atomic?